Types for existing npm package `subleveldown` with tests

carsonfarmer opened the pull request:
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

demensky commented on types/subleveldown/index.d.ts:
```diff
@@ -0,0 +1,33 @@
+declare namespace sub {
+    interface Options {
+        separator?: string;
+        open?: ((callback?: ErrorCallback) => any);
```
Please Use `void` as [return type of callback][callback].

[callback]: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#return-types-of-callbacks

carsonfarmer replied:
Done

demensky commented on types/subleveldown/index.d.ts:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+declare function sub<DB, Iterator>(db: LevelUp < DB, Iterator >, prefix ?: string, options ?: sub.Options): sub.SublevelDown<DB, Iterator>;
```
[`opts` may be a `string`](https://github.com/Level/subleveldown/blob/master/leveldown.js#L48).

carsonfarmer replied:
Good catch, added!

demensky commented on types/subleveldown/index.d.ts:
```diff
@@ -0,0 +1,33 @@
+    interface Options {
+        separator?: string;
+        open?: ((callback?: ErrorCallback) => any);
+        [key: string]: any;
```
The source code uses only [separator] and [open]. Why is it here?

[separator]: https://github.com/Level/subleveldown/blob/master/leveldown.js#L51

[open]: https://github.com/Level/subleveldown/blob/master/leveldown.js#L62

carsonfarmer replied:
That was a relic of some old work, thanks for catching! Fixed.

demensky commented on types/subleveldown/index.d.ts:
```diff
@@ -0,0 +1,33 @@
+declare namespace sub {
+    interface Options {
+        separator?: string;
+        open?: ((callback?: ErrorCallback) => any);
```
[Why is callback listed as optional?][optional-parameters-in-callbacks] In [source code][_open], it is always passed.

[optional-parameters-in-callbacks]: https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#optional-parameters-in-callbacks
[_open]: https://github.com/Level/abstract-leveldown/blob/52d0179e272f2f5c0420f3e24415510aa63e042b/abstract-leveldown.js#L34

carsonfarmer replied:
Good point, and thanks for the link! Fixed.

demensky commented on types/subleveldown/index.d.ts:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+declare function sub<DB, Iterator>(db: LevelUp < DB, Iterator >, prefix ?: string, options ?: sub.Options): sub.SublevelDown<DB, Iterator>;
```
`SubIterator` is actually a class. I suggest using functions and ambient classes. [Example].

[Example]: https://www.typescriptlang.org/play/#code/MYGwhgzhAECCBGEAuAnMwkEkkFM1IHsUAeAEQCEAaabPMQlAPmgG8BfAWAChuBLAO1woAZuhzQAMjgBuOEAFUADmSo0h9Is3Z9BeUcHEB5RUl4F+MbTy4B6G9AC0T5y4fduAExygwKcT6hoAGUAV3hafCIVagiNJmgcAA9cfg8YBGQ0DFiGaLU6BkYWbmhS6CQAW0UALmgwfgBPAG53LjLoRTCQXmBoYHNMkIwiAAoPeFqpWQVFaDyczWpFP2FeRIB+WsyBAHNqAhMzC2hN6GNTAYBKFq5Oay8fP2hhEP4MI+CwhZIKGPVCsYTSQyORKeb-RYdFZrU7bfh7aAHC4WU7nI4QS61ULhCE-VTfRg3bj9CxIco4ZAARmgAF5oPwcAB3T44gqjdjXaB2FnfVok5Dk5AAJlpPNxIw5TS59mxvOsuCpADpKooiVwFUghcqqk0gA

carsonfarmer replied:
This sounds reasonable. In this case, do you mean the main `SubDown` function/class, or the 'internal' `SubIterator`? In the source code, `SubIterator` is not exported, though it obviously is returned from the `iterator` method. Are you suggesting:
- [ ] Add this `SubIterator` class to the definition as well?
- [ ] Simply implement the main `SubDown` using a function and ambient class?
- [ ] Both of these?

Note that I updated the name of the main interface (`SubDown`) to better match the source code.

carsonfarmer replied:
Either way, your suggestion is a good one for the main class/function/interface. So I will certainly add this.

carsonfarmer commented:
Thank you @demensky for your thorough and helpful review! I've addressed most of your comments, and have one open question left for some clarification.

carsonfarmer commented on types/subleveldown/index.d.ts:
```diff
@@ -4,27 +4,18 @@
-    interface SubDown<DB, Iterator> extends LevelUp<DB, Iterator> {
-        clear: InferDBClear<DB>;
-    }
+    interface SubDown<DB, Iterator> extends LevelUp<DB, Iterator> {}
```
Removed clear definitions in favour of a separate PR to levelup (where they actually belong): https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40792

carsonfarmer commented:
More updates, I'm still new to some aspects of Typescript defs, so I hope this is now pretty clear.

The most recent updates allow the types to reflect arbitrary key/value types in each sublevel, which is an explicit requirement of the package: https://github.com/Level/subleveldown#background. The tests have been updated to explicitly check for this.

carsonfarmer commented on types/subleveldown/index.d.ts:
```diff
@@ -0,0 +1,26 @@
+import { LevelUp } from 'levelup';
+import { AbstractLevelDOWN, AbstractIterator, ErrorCallback } from 'abstract-leveldown';
+
+declare namespace sub {
```
Is this the correct use of a namespace here?

carsonfarmer replied:
As in, should this be named `SubDown`?

demensky replied:
I think not.  In the [source code](https://github.com/Level/subleveldown/blob/master/index.js#L5), a function is exported, not a class. The class instance only returns from the function.

I probably made some unintentional mistakes in some comments because I got confused in the source code of the library.

I can review this PR in more detail and thoughtfully. What do you say about this?

carsonfarmer replied:
Yes please, and not a problem at all!

demensky replied:
I suggest such improvements: carsonfarmer/DefinitelyTyped#1.

carsonfarmer commented:
Thanks again to @demensky! I look forward to your (hopefully final) review, and am happy to make any additional changes as needed. I suspect we can actually simplify things a great deal here.

carsonfarmer commented:
Apologies to anyone following this work, I accidentally pushed a ton of commits to this branch. I've now force-pushed the the branch from 98f2143 to 3ba1bbe... hopefully this doesn't cause you any issues!

carsonfarmer commented:
Ok, I'm happy with the final product here @demensky! I've made one last commit to add your name to the header, I hope this is ok? Otherwise, I think we're good to go here?

carsonfarmer commented:
I think this one is ready to go. It seems like the Revision needed label got left on for some reason?
